logictest: deflake partial_stats

yuzefovich opened the pull request:
Whenever we collect a new statistic, in order to guarantee that it'll be observed by the optimizer on the following query, we must clear the stats cache after having collected that stat. This `Clear` call was missing after having collected the partial stat (which also results in generating the merged stat). It was incompletely fixed in 5707cf173b205e95f0a77ad15284c60f488977a2 by adding a retry to the EXPLAIN and resulted in a rare case where the merged stat is not observed by a later query. (I'm not quite sure why `retry` on an earlier statement is insufficient to guarantee that the merged stat is observed by a later statement, given that no more new stats are created. I did see the flake go away with this patch.) This commit clears the stats cache addressing the flake. (Also it looks like this flake was exposed via cabf4826c4fb2067d6e26581ce679c1ef4465cf2.)

Having a `retry` on the EXPLAIN query is an alternative way of fixing the flake. Some more context can be found at 5707cf173b205e95f0a77ad15284c60f488977a2.

Fixes: #149309.

Release note: None

yuzefovich commented:
blathers auth-check

yuzefovich commented:
You shall pass

yuzefovich commented:
(Debugging some blathers stuff [slack](https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1751923396697579))

michae2 reviewed (commented):
There are three code paths where we pick up new stats:

1. [getTableStatsFromCache](https://github.com/cockroachdb/cockroach/blob/00ecaf44a1e091dafa3176f64edbca62dda65663/pkg/sql/stats/stats_cache.go#L346) when the stats are not in the cache (or are stale)
2. [refreshTableStats](https://github.com/cockroachdb/cockroach/blob/00ecaf44a1e091dafa3176f64edbca62dda65663/pkg/sql/stats/stats_cache.go#L546) when an event on system.table_statistics comes from the rangefeed
3. [ShowTableStats](https://github.com/cockroachdb/cockroach/blob/00ecaf44a1e091dafa3176f64edbca62dda65663/pkg/sql/show_stats.go#L67) during SHOW

Clearing the stats cache exercises code path (1). Using `retry` exercises code path (2). It'd be nice to keep some of the retries so that we still have some coverage of code path (2). Maybe we could keep the retries in `distsql_stats`?

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149504#-:-OUarJDu5lnCc5tMYwY7:b8zl781)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
logictest: deflake partial_stats

Whenever we collect a new statistic, in order to guarantee that it'll be
observed by the optimizer on the following query, we must clear the
stats cache after having collected that stat. This `Clear` call was
missing after having collected the partial stat (which also results in
generating the merged stat). It was incompletely fixed in
5707cf173b205e95f0a77ad15284c60f488977a2 by adding a retry to the
EXPLAIN and resulted in a rare case where the merged stat is not
observed by a later query. (I'm not quite sure why `retry` on an earlier
statement is insufficient to guarantee that the merged stat is observed
by a later statement, given that no more new stats are created. I did
see the flake go away with this patch.) This commit clears the stats cache
addressing the flake. (Also it looks like this flake was exposed via
recent cabf4826c4fb2067d6e26581ce679c1ef4465cf2.)

Having a `retry` on the EXPLAIN query is an alternative way of fixing
the flake. Some more context can be found at
5707cf173b205e95f0a77ad15284c60f488977a2.

Release note: None

yuzefovich reviewed (commented):
Good point - seems reasonable to me, so I reverted the changes from `distsql_stats`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149504#-:-OUbX1oD9i28skQ7OUYZ:b-wiyah)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149504#-:-OUcJysp1GlNN1wETMaf:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->
